saleor/graphql/product/resolvers.py

```diff
@@ -47,20 +47,28 @@ def resolve_digital_contents(_info):
     return models.DigitalContent.objects.all()
 
 
-def resolve_product_by_id(info, id, channel_slug, requestor):
-    return (
-        models.Product.objects.visible_to_user(requestor, channel_slug=channel_slug)
-        .filter(id=id)
-        .first()
-    )
+def resolve_product_by_id(
+    info, id, channel_slug, requestor, requestor_has_access_to_all
+):
+    visible_products = models.Product.objects.visible_to_user(requestor, channel_slug)
+    if not requestor_has_access_to_all:
+        visible_products = visible_products.annotate_visible_in_listings(
+            channel_slug
+        ).exclude(visible_in_listings=False)
```

Comment:
Maybe we should make a separate function from these lines?

We have the same block of code in many resolvers.

Reply:
I will do that